Rewrite dotnet-getdocument and GetDocument.Insider to retrieve all documents in one go

dougbu opened the pull request:
- #8242 1 of 2
- save a cache file listing all retrieved documents
- remove .NET Core App 2.0 support
- remove ServiceProjectReferenceMetadata.targets, related `Project` class, and searches for a project
  - tools will run within a project and get needed information from project on command line
- roll framework forward in both tools to expand their applicability

nits:
- refactor methods in `GetDocumentCommandWorker`
- reorder option addition for consistency and to place `--help` at the top of help output
- consolidate information about method signatures at top of `GetDocumentCommandWorker`
- consolidate `try` / `catch` blocks in `GetDocumentCommandWorker`

dougbu committed:
Rewrite dotnet-getdocument and GetDocumentCommandInsider to retrieve all documents in one go
- #8242 1 of 2
- save a cache file listing all retrieved documents
- remove .NET Core App 2.0 support
- remove ServiceProjectReferenceMetadata.targets, related `Project` class, and searches for a project
  - tools will run within a project and get needed information from project on command line
- roll framework forward in both tools to expand their applicability

nits:
- refactor methods in `GetDocumentCommandWorker`
- reorder option addition for consistency and to place `--help` at the top of help output
- consolidate information about method signatures at top of `GetDocumentCommandWorker`
- consolidate `try` / `catch` blocks in `GetDocumentCommandWorker`

rynowak commented on src/Mvc/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs:
```diff
@@ -22,200 +46,325 @@ public static int Process(GetDocumentCommandContext context)
 
-            var services = GetServices(entryPointType, context.AssemblyPath, context.AssemblyName);
-            if (services == null)
+            static string FormatException(Exception exception) => $"{exception.GetType().FullName}: {exception.Message}";
```
consider putting this after the return statement to separate it a bit. It's a little jarring to see a single-line local function defined in the middle of a method.

rynowak replied:
I would really really strongly encourage the use of `Exception.ToString()` instead of hiding the information behind a verbose logging switch. 

By doing this you can optimize for users getting an experience that they troubleshoot themselves (failure in their own app), or can use to immediately report a problem to us (copy-paste the call stack).

javiercn reviewed (approved):
Pranav asked me to help him look at this.

Overall I think looks good. I left a few comments. Mostly I would strongly recommend you to reuse the logic for finding the app entry point, as we share that everywhere.

I'm not sure about the constraints this tool has, but there are other things that are a bit concerning to me, like reading the deps file or the runtime.config.json file.

I also don't see any tests for this PR, not sure if its because there are already tests and didn't require changes or because there aren't any. If its the latter, I think there should be at least one, as this tool does a lot of things that are prone to integration errors.

That said, I don't see anything that strikes me as wrong here.

rynowak commented on src/Mvc/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs:
```diff
@@ -22,200 +46,325 @@ public static int Process(GetDocumentCommandContext context)
+            }
+
+            return builder.Build().Services;
+        }
```
/cc @javiercn

rynowak commented on src/Mvc/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs:
```diff
@@ -22,200 +46,325 @@ public static int Process(GetDocumentCommandContext context)
+            var stream = File.Create(context.FileListPath);
+            using var writer = new StreamWriter(stream) { AutoFlush = true };
+            writer.WriteLine(string.Join(Environment.NewLine, filePathList));
+            stream.Flush();
```
Is there a really strong need for either `AutoFlush` or the explict call to `flush`? the stream will flush when it's disposed.

rynowak replied:
I've noticed that this is really prevalent all over the code, I'd be interested to learn if this is here for any important reason.

rynowak commented on src/Mvc/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs:
```diff
@@ -22,200 +46,325 @@ public static int Process(GetDocumentCommandContext context)
+            Reporter.WriteInformation(Resources.FormatRetrievingDocument(documentName));
+
+            var stream = new MemoryStream();
+            using var writer = new StreamWriter(stream) { AutoFlush = true };
```
Similar question about `AutoFlush`

rynowak commented on src/Mvc/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs:
```diff
@@ -22,200 +46,325 @@ public static int Process(GetDocumentCommandContext context)
+                return null;
+            }
+
+            var finishedIndex = Task.WaitAny(resultTask, Task.Delay(TimeSpan.FromMinutes(1)));
```
Could this be simplified using https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.wait?view=netframework-4.8#System_Threading_Tasks_Task_Wait_System_TimeSpan_

rynowak commented on src/Mvc/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs:
```diff
@@ -22,200 +46,325 @@ public static int Process(GetDocumentCommandContext context)
+                        DocumentService,
+                        documentName));
+
+                    return null;
```
Is there a reason for this to be inside the `try/catch`? This seems like it belongs higher up.

rynowak commented on src/Mvc/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs:
```diff
@@ -22,200 +46,325 @@ public static int Process(GetDocumentCommandContext context)
-                    Reporter.WriteWarning(FormatException(innerException));
+                    sanitizedDocumentName = string.Join(
+                        ".",
+                        sanitizedDocumentName.Split(InvalidFilenameStrings, StringSplitOptions.None));
```
Could this simplified and and made more obvious using `string.Replace`?

rynowak commented on src/Mvc/dotnet-getdocument/src/Commands/InvokeCommand.cs:
```diff
@@ -16,155 +16,123 @@ internal class InvokeCommand : HelpCommandBase
-                args.Add(project.TargetPath);
+                args.Add(assemblyPath);
+                args.Add("--project");
+                args.Add(projectName);
```
I'm probably not the best person to review the inside-man stuff. I've never really known how this works 😕

rynowak commented:
The general spirit of the changes here makes sense to me and sees like a good enhancement. Due to the diffs being what they are, I'm not really sure if I'm reviewing old code or new code, but I still found some things that seem worthy of a look.

dougbu commented:
@javiercn I realize I missed your high-level comments, sorry:
>  things that are a bit concerning to me, like reading the deps file or the runtime.config.json file.

None of the code reads either of these files though it does scan the project assets file for probing paths. The `dotnet` command line doesn't have a `--please-also-use-these-assets` option.

See similar code in the `dotnet-ef` tool: https://github.com/aspnet/EntityFrameworkCore/blob/078449fd1363bc3d79778a4f726262e2ced30ae5/src/dotnet-ef/RootCommand.cs#L127-L139 So, fairly battle-tested

> I also don't see any tests for this PR

This feature has some engineering debt I'll be catching up on in the next milestone. #4914 covers the test part of that.

dougbu committed:
Address PR comments
- use Microsoft.Extensions.HostFactoryResolver.Sources
  - remove `GetMethod(...)` parameters that aren't needed anymore
  - remove Microsoft.AspNetCore.Hosting.Abstractions dependency
- remove `Flush()` and `AutoFlush = true`
- use `Exception.ToString()` and `String.Replace(...)`
- move unnecessary lines out of a `try` block

nits:
- "retrieving" -> "generating"
- shorten the lifespan of a `Task`
- ensure GetDocument.Insider exit codes are unique
- make a few more `string`s `const`
- fold a few expressions over fewer lines

dougbu committed:
!fixup! Shorten lifetime of a `StreamWriter`
- ensure it is flushed before underlying stream is read

dougbu reviewed (commented):
:up::date; to address more comments than I expected

dougbu commented on src/Mvc/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs:
```diff
@@ -22,200 +46,325 @@ public static int Process(GetDocumentCommandContext context)
+            var stream = File.Create(context.FileListPath);
+            using var writer = new StreamWriter(stream) { AutoFlush = true };
+            writer.WriteLine(string.Join(Environment.NewLine, filePathList));
+            stream.Flush();
```
FYI I had to do this with the other `StreamWriter` because the `MemoryStream` was used before it was disposed. Did the more obvious thing of giving those two objects separate lifetimes. Cleaned this instance up.
